### PR TITLE
rpc: add missing fields in send_transaction

### LINKED ISSUE #1909: Fill missing fields TransactionRequest fields
### Describe the feature

In https://github.com/paradigmxyz/reth/pull/1774 we added `send_transaction` impl.

we're missing some updates after we converted the request to properly fill in missing fields here:

https://github.com/paradigmxyz/reth/blob/5987ecb1949604b0e8acb5fa18ec0591460aca0a/crates/rpc/rpc/src/eth/api/transactions.rs#LL175

Basically same as:

https://github.com/foundry-rs/foundry/blob/da2392e58bb8a7fefeba46b40c4df1afad8ccd22/anvil/src/eth/api.rs#L2209-L2236

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bharath-123**
resolves: https://github.com/paradigmxyz/reth/issues/1909 and https://github.com/paradigmxyz/reth/issues/1978

### CODE REVIEW

**Comment by bharath-123** [comment]
Hey @mattsse my first PR for reth, tried to make self.get_chain_id work by constrainting the trait bounds other Network. Do let me know if this is acceptable or not

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 1]
```diff
@@ -192,14 +193,51 @@ where
...
+        let gas_price = request.gas_price.unwrap_or_default();
+        let max_fee_per_gas = request.max_fee_per_gas.unwrap_or_default();
+
         let transaction = match request.into_typed_request() {
-            Some(tx) => tx,
+            Some(TypedTransactionRequest::Legacy(mut m)) => {
+                // nonce was not specified, we now need to update it to the latest nonce
+                if m.nonce.eq(&U256::ZERO) {
+                    m.nonce = nonce
+                }
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/transactions.rs:210]: this we need to set beforehand, before building

just opened which is relevant here  https://github.com/paradigmxyz/reth/issues/1978

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 2]
```diff
@@ -192,14 +193,51 @@ where
...
             None => return Err(SignError::NoAccount.into()),
         };
+
+        let chain_id = self.chain_id();
+        let nonce = self.get_transaction_count(from, None)?;
+        // we might need an oracle to get the gas limit of the current n/w as well if i m right?
+        let gas_limit = request.gas.unwrap_or_default();
+        // we need an oracle to fetch the gas price of the current chain
+        let gas_price = request.gas_price.unwrap_or_default();
+        let max_fee_per_gas = request.max_fee_per_gas.unwrap_or_default();
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/transactions.rs:203]:
```suggestion
        // TODO add defaults via fee oracle <https://github.com/paradigmxyz/reth/issues/1121>
        let gas_price = request.gas_price.unwrap_or_default();
        let max_fee_per_gas = request.max_fee_per_gas.unwrap_or_default();

        // TODO estimate via estimate function
        let gas_limit = request.gas.unwrap_or_default();
```

**Review by mattsse — CHANGES_REQUESTED** [review]
looks great,

smol nit re nonce, and TODO comments.

also for the gas_limit, we should estimate gas by calling `estimate_gas` if not set.

File: crates/rpc/rpc/src/eth/api/transactions.rs (hunk 1 above)
Comment (bharath-123) [reply crates/rpc/rpc/src/eth/api/transactions.rs:210]: you mean before building the tx type right?

**Comment by bharath-123** [comment]
I figured that we can use estimate gas but didn't go that route since I thought that we would estimate the gas, populate the request fields and call the send_transaction. Thanks for clarifying this @mattsse

File: crates/rpc/rpc/src/eth/api/transactions.rs (hunk 1 above)
Comment (mattsse) [reply crates/rpc/rpc/src/eth/api/transactions.rs:210]: yes because this setting is universal

**Comment by bharath-123** [comment]
got it @mattsse this is clear to me!! Complete PR coming your way in the next 1-2 hours!

**Comment by bharath-123** [comment]
hey @mattsse could you give it a pass?

File: crates/rpc/rpc/src/eth/api/state.rs [hunk 3]
```diff
@@ -31,9 +33,32 @@ where
...
+                    }
+                });
+
+                let on_chain_nonce = if let Some(highest_nonce_tx) = highest_nonce {
+                    U256::from(highest_nonce_tx.transaction.nonce())
+                } else {
+                    U256::from(0)
+                };
+
+                EthResult::Ok(on_chain_nonce)
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/state.rs:55]: this needs to be an if let and should not return if `address_txs` is empty, instead it should go into the other arm.

File: crates/rpc/rpc/src/eth/api/state.rs [hunk 4]
```diff
@@ -58,7 +83,7 @@ where
 
         // TODO: remove when HistoricalStateProviderRef::proof is implemented
         if !block_id.is_latest() {
-            return Err(EthApiError::InvalidBlockRange)
+            return Err(EthApiError::InvalidBlockRange);
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/state.rs:86]: please use nightly formatting

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 5]
```diff
@@ -187,19 +188,69 @@ where
...
         let from = match request.from {
             Some(from) => from,
             None => return Err(SignError::NoAccount.into()),
         };
+
+        // set nonce if not already set before
+        if request.nonce.is_none() {
+            let nonce =
+                self.get_transaction_count(from, Some(BlockId::Number(BlockNumberOrTag::Pending)))?;
+            request.nonce = Some(nonce.saturating_add(U256::from(1)))
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/transactions.rs:201]: this should be checked_add and map error to `MaxNonce` see `EthApiError`

File: crates/transaction-pool/src/lib.rs [hunk 6]
```diff
@@ -325,6 +325,20 @@ where
     fn on_propagated(&self, txs: PropagatedTransactions) {
         self.inner().on_propagated(txs)
     }
+
+    fn get_all_transactions_by_sender(
```
Comment (mattsse) [inline crates/transaction-pool/src/lib.rs:329]: we need the same thing in the other layers as well, see `fn pooled_transactions` for example.

File: crates/transaction-pool/src/traits.rs [hunk 7]
```diff
@@ -145,6 +145,9 @@ pub trait TransactionPool: Send + Sync + Clone {
     ///
     /// Consumer: P2P
     fn on_propagated(&self, txs: PropagatedTransactions);
+
+    /// Returns all transactions sent by a given user
+    fn get_all_transactions_by_sender(&self, sender: Address) -> Vec<Arc<ValidPoolTransaction<Self::Transaction>>>;
```
Comment (mattsse) [inline crates/transaction-pool/src/traits.rs:150]:
```suggestion
    fn get_transactions_by_sender(&self, sender: Address) -> Vec<Arc<ValidPoolTransaction<Self::Transaction>>>;
```

**Review by mattsse — CHANGES_REQUESTED** [review]
nice work, 
just a few nits.

File: crates/rpc/rpc/src/eth/api/transactions.rs (hunk 5 above)
Comment (bharath-123) [reply crates/rpc/rpc/src/eth/api/transactions.rs:201]: yes

File: crates/transaction-pool/src/lib.rs (hunk 6 above)
Comment (bharath-123) [reply crates/transaction-pool/src/lib.rs:329]: will add ser!

**Comment by bharath-123** [comment]
hey @mattsse could you give this a go when you are free? :)

**Review by mattsse — APPROVED** [review]
ty, lgtm.

only did one change: move +1 to the `transaction_count` function because this is the next available nonce

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 8]
```diff
@@ -224,19 +225,69 @@ where
...
+                    to: request.to,
+                    gas: request.gas,
+                    gas_price: Some(U256::from(gas_price)),
+                    max_fee_per_gas: Some(U256::from(max_fee_per_gas)),
+                    value: request.value,
+                    data: request.data.clone(),
+                    nonce: request.nonce,
+                    chain_id: Some(chain_id),
+                    access_list: request.access_list.clone(),
+                    max_priority_fee_per_gas: Some(U256::from(max_fee_per_gas)),
```
Comment (huongcodelight) [inline crates/rpc/rpc/src/eth/api/transactions.rs:259]: I try to run this method and the underlying `ensure_fees` function return error `Err(EthApiError::ConflictingFeeFieldsInRequest)`. I debugged and found that there is no case that all of three `gas_price`, ` max_fee_per_gas` and `  max_priority_fee_per_gas` is Some(...). Could anyone please help me with this point ?

Comment (mattsse) [reply crates/rpc/rpc/src/eth/api/transactions.rs:259]: could you please open a new issue for this?

**Pull request merged**